### DISCUSSION TITLE
Can't the url with custom-protocol be strictly same in different OS?

### DISCUSSION THREAD

**Discussion by percy507** [tauri-apps/tauri#5506] [Q&A] [answered]
Imagine a situation, when using rich-text editor, we usually save its content as html string. The content may contains img which `src` is an url with custom-protocol. 

We suppose current OS is macOS and the img's src  is `custom://localhost/dog.png`. Then we sync the content to windows OS, I wonder if the img's src(`custom://localhost/dog.png`)  still work? If doesn't work, except converting `custom://localhost/dog.png` into `https://custom.localhost/dog.png` manually, is there any better way to handle this?

**Comment by FabianLars** [comment, marked as answer]
Nope, Windows doesn't support custom schemes yet, but we do have https://tauri.app/v1/api/js/tauri/#convertfilesrc for this (which tbh is not a solution if you wanna hardcode the path in your html)

**Reply by percy507** [reply]
Thanks for your reply. I'll think about how to solve this.

**Reply by JonasKruckenberg** [reply]
Honestly, you should just omit the base and store relative paths in your html if possible, that way your documents will be portable even when you decide to add non-Tauri clients (like a web-based viewer or something)

**Reply by percy507** [reply]
> Honestly, you should just omit the base and store relative paths in your html if possible, that way your documents will be portable even when you decide to add non-Tauri clients (like a web-based viewer or something)

Of course I can store relative path. But then how can I resolve its content and do some extra works  (eg: if the content is encrypted, I also need to decrypt it)? 

```bash
# development mode
http://localhost:6777/dog.png

# release mode
tauri://localhost/dog.png
```

**Reply by percy507** [reply]
Current custom protocol may like this.

```bash
# linux, macos
custom://localhost/dog.png

# windows
https://custom.localhost/dog.png
```

Consider of windows doesn't support custom scheme yet, what about providing an option to let user decide to use native custom-protocol or windows-like custom-protocol on linux/macos when registering custom-protocol? Then the url with custom-protocol can be same in different OS.

> ps: I don't know if it is easy to add the option. Just a thought. 😂

**Reply by FabianLars** [reply]
I actually looked into that because of the cookie issues, hoping that it's only related to the scheme which means that switching to http(s) could fix it.
Buuut the api we use on Windows doesn't exist on macOS/Linux and i didn't find an alternative :/

**Reply by percy507** [reply]
> Notice this issue https://github.com/tauri-apps/tauri/issues/4899

I have another thought. Maybe we can let user store the custom-protocol url in windows-style (`https://$SCHEME.localhost/$PATH`) for cross-platform, but in macOS we intercept the request and rewrite it to macos-style (`$SCHEME://localhost/$PATH`) programmatically.

**Reply by FabianLars** [reply]
Also not possible because we can't rewrite the request like we can on Windows (btw this is how we simulate the custom protocol on Windows). So we can't intercept it on the "native" side, and injecting JavaScript to intercept it on the js side (idk if that would even work) is a bit too unreliable imho.

**Reply by percy507** [reply]
Lol, I choose the simple way😂. Before render the html string on macOS, find the urls in windows-style and convert them to macOS-style and vice versa.

**Reply by FabianLars** [reply]
Yeah we can't rely on that idea on our side unfortunately, but since it's now documented here devs could do that on their side if they want 💪
